rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::datafusion::LanceTableProvider;
+use crate::Dataset;
+use arrow_array::{Array, RecordBatch, StringArray};
+use datafusion::dataframe::DataFrame;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::prelude::SessionContext;
+use std::sync::Arc;
+
+/// A SQL builder to prepare options for running SQL queries against a Lance dataset.
+#[derive(Clone, Debug)]
+pub struct SqlBuilder {
+    /// The dataset to run the SQL query
+    pub(crate) dataset: Dataset,
+
+    /// The SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: Option<String>,
+
+    /// If true, the query result will include the internal row id
+    pub(crate) row_id: Option<bool>,
+
+    /// If true, the query result will include the internal row address
+    pub(crate) row_addr: Option<bool>,
+}
+
+impl SqlBuilder {
+    pub fn new(dataset: Dataset, sql: &str) -> Self {
+        Self {
+            dataset,
+            sql: sql.to_string(),
+            table_name: Some("dataset".to_string()),
+            row_id: None,
+            row_addr: None,
+        }
+    }
+
+    /// The table name to register in the datafusion context.
+    /// This is used to specify a "table name" for the dataset.
+    /// So that you can run SQL queries against it.
+    /// If not set, the default table name is "dataset".
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = Some(table_name.to_string());
+        self
+    }
+
+    /// Specify if the query result should include the internal row id.
+    /// If true, the query result will include an additional column named "_rowid".
+    pub fn with_row_id(mut self, row_id: bool) -> Self {
+        self.row_id = Some(row_id);
+        self
+    }
+
+    /// Specify if the query result should include the internal row address.
+    /// If true, the query result will include an additional column named "_rowaddr".
+    pub fn with_row_addr(mut self, row_addr: bool) -> Self {
+        self.row_addr = Some(row_addr);
+        self
+    }
+
+    pub async fn build(self) -> lance_core::Result<SqlQueryBuilder> {
+        let ctx = SessionContext::new();
+        let row_id = self.row_id.unwrap_or(false);
+        let row_addr = self.row_addr.unwrap_or(false);
+        ctx.register_table(
+            self.table_name.unwrap(),
+            Arc::new(LanceTableProvider::new(
+                Arc::new(self.dataset.clone()),
+                row_id,
+                row_addr,
+            )),
+        )?;
+        let df = ctx.sql(&self.sql).await?;
+        Ok(SqlQueryBuilder::new(df))
+    }
+}
+
+pub struct SqlQueryBuilder {
```

Comment:
sorry maybe I was not clear, I was saying the `SqlBuilder` could be renamed as `SqlQueryBuilder` to match the result of `build` that is a `SqlQuery`, so this struct should still be named `SqlQuery`.